proposals/partial-type-inference.md

```diff
@@ -0,0 +1,738 @@
+  > ```
+
+  An existing solution can be seen in `Create()` method wrappers of constructors enabling a type inference through method type inference as you can se in the examples above.
+  However, we can't use it with *object_or_collection_initializer*; we are limited by method type inference strength, and it adds unnecessary boiler code.
+
+  Adding constructor type inference as we will describe in the following section would solve above mentioned examples.
+
+  > Example
+  >
+  > ```csharp
+  > var wrappedData = new Wrapper<>(new MyData);
+  > class Wrapper<T> { public Wrapper(T item) { ... } }
+  > ```
+  > 
+  > ```csharp
+  > var alg = new Algorithm<>(new MyData());
+  > var algWithSpecialLogger = new Algorithm<_ , SpecialLogger<_>>(new MyData());
+  > 
+  > class Algorithm<TData, TLogger> where TLogger : Logger<TData> { public Algorithm(TData data) { ... }}
+  > ```
+  > 
+  > ```csharp
+  > var numbers = new List<>() {1,2,3,4};
+  > ```
+  > 
+  > ```csharp
+  > new Dictionary<_, string>() {[""] = null};
+  > ```
+
+No matter how the partial type inference would work, we should be careful about the following things.
+
+- **Convenience** - We want an easy and intuitive syntax that we can skip the obvious type arguments.
+- **Performance** - Type inference is a complicated problem when we introduce subtyping and overloading in a type system.
+Although it can be done, the computation can take exponential time which we don't want.
+So it has to be restricted to cases, where the problem can be solved effectively but it still has practical usage.
+- **IDE** - Improvement of the type inference can complicate IDE hints during coding. 
+We should give the user clear and not overwhelming errors when there will be an error and try to provide info that helps him to fix it.
+- **Extensions** - We don't want to make this change blocker for another potential feature in the future. 
+So will want to look ahead to other potential directions, which can be done after this feature.
+
+## Detailed design
+[design]: #detailed-design
+
+### Grammar
+
+> Specification: Original section changed in the following way
+
+The following changes are made in [tokens](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/lexical-structure.md#64-tokens) located in the [grammar](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/lexical-structure.md#62-grammars) section.
+
+> [Identifiers](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/lexical-structure.md#643-identifiers)
+
+- The semantics of an identifier named `_` depends on the context in which it appears:
+  - It can denote a named program element, such as a variable, class, or method, or
+  - It can denote a discard (§9.2.9.1).
+  - \***It can denote an inferred type argument avoiding specifying type arguments which can be inferred by the compiler.**
```

Comment:
Rather than _it can_ I'd prefer the language _it will_. Essentially when flipping to `/langversion:X` where `X` is the release that introduces this feature, I'd prefer the meaning to be unambiguous. That is for the case `M<_, object>(42, null)`, the `_` unambiguously means an inferred type parameter. 

Yes this would be breaking change to developers who defined `class _ { }` but that seems like a minority case. It in the same ballpark as `class record` which we broke when introducing records.

Reply:
Should be fixed by commit [33b71f0](https://github.com/dotnet/csharplang/pull/7582/commits/33b71f0c489eaa03018d18a9cabf8b2cdf74a43a)